### PR TITLE
663/mk/standardize op client requests

### PR DESCRIPTION
**Pull request by mkurapov**
## Changes proposed in this pull request

- Update incoming payment & outgoing payment client `list` `create` `complete` requests to use `paymentPointer` as an argument instead of `url`, e.g.:

```ts
await client.incomingPayment.create({ url, accessToken, body: { description, incomingAmount } } // before
await client.incomingPayment.create({ paymentPointer, accessToken }, { description, incomingAmount }) // after
```
- Update tests to get 100% coverage:
Screen Shot 2023-01-13 at 1 57 29 PM

## Context

- Parent issue: #663

## Checklist

- [x] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [ ] Make sure that all checks pass

### CODE REVIEW

File: packages/open-payments/src/client/ilp-stream-connection.test.ts [hunk 1]
```diff
@@ -15,14 +28,34 @@ describe('ilp-stream-connection', (): void => {
...
+
+    describe('get', (): void => {
+      test('calls get method with correct validator', async (): Promise<void> => {
+        const mockResponseValidator = ({ path, method }) =>
+          path === '/connections/{id}' && method === HttpMethod.GET
+
+        jest
+          .spyOn(openApi, 'createResponseValidator')
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          .mockImplementation(mockResponseValidator as any)
```
Comment (wilsonianb) [inline packages/open-payments/src/client/ilp-stream-connection.test.ts:42]: Could we check that the spy is called?

File: packages/open-payments/src/client/incoming-payment.ts [hunk 2]
```diff
@@ -9,28 +9,25 @@ import {
...
 }
 
-interface PostArgs<T = undefined> {
-  url: string
-  body?: T
-  accessToken: string
-}
-
-interface ListGetArgs {
+interface RequestWithPaymentPointerArgs {
```
Comment (wilsonianb) [inline packages/open-payments/src/client/incoming-payment.ts:17]: What do you think of `(Sub)ResourceRequestArgs` and `CollectionRequestArgs`?

That might also somewhat future-proof this for
- https://github.com/interledger/open-payments/issues/206

File: packages/open-payments/src/client/incoming-payment.test.ts [hunk 3]
```diff
@@ -42,83 +42,21 @@ describe('incoming-payment', (): void => {
 
   const axiosInstance = defaultAxiosInstance
   const logger = silentLogger
-  const baseUrl = 'http://localhost:1000'
+  const paymentPointer = `http://localhost:1000/.well-known/pay`
```
Comment (wilsonianb) [inline packages/open-payments/src/client/incoming-payment.test.ts:45]:
```suggestion
  const paymentPointer = `http://localhost:1000/.well-known/pay`
  const accessToken = 'accessToken'
```

File: packages/open-payments/src/client/incoming-payment.test.ts [hunk 4]
```diff
@@ -140,7 +78,9 @@ describe('incoming-payment', (): void => {
         }
       })
 
-      nock(baseUrl).get('/incoming-payments').reply(200, incomingPayment)
+      nock(paymentPointer)
+        .get('/incoming-payments/1')
+        .reply(200, incomingPayment)
 
       await expect(() =>
```
Comment (wilsonianb) [inline packages/open-payments/src/client/incoming-payment.test.ts:85]:
```suggestion
      await expect(
```

File: packages/open-payments/src/client/outgoing-payment.test.ts [hunk 5]
```diff
@@ -28,71 +37,27 @@ describe('outgoing-payment', (): void => {
 
   const axiosInstance = defaultAxiosInstance
   const logger = silentLogger
-  const baseUrl = 'http://localhost:1000'
+  const paymentPointer = `http://localhost:1000/.well-known/pay`
   const openApiValidators = mockOpenApiResponseValidators()
 
-  describe('createOutgoingPaymentRoutes', (): void => {
```
Comment (mkurapov) [inline packages/open-payments/src/client/outgoing-payment.test.ts:34]: not needed, since `routes` describe block takes care of this

File: packages/open-payments/src/test/helpers.ts [hunk 6]
```diff
@@ -31,6 +31,23 @@ export const defaultAxiosInstance = createAxiosInstance({
...
+  return async () => {
+    const savedEnvVars = Object.assign({}, process.env)
+
+    for (const key in override) {
+      process.env[key] = override[key]
+    }
+
+    await testCallback()
+
+    process.env = savedEnvVars
```
Comment (wilsonianb) [inline packages/open-payments/src/test/helpers.ts:47]: Should this be in a `try`/`finally`?

File: packages/open-payments/src/test/helpers.ts [hunk 7]
```diff
@@ -39,6 +41,26 @@ export const mockOpenApiResponseValidators = () => ({
...
+  kty: 'OKP',
+  crv: 'Ed25519',
+  ...overrides
+})
+
+export const mockPaymentPointer = (
+  overrides?: Partial<PaymentPointer>
+): PaymentPointer => ({
+  id: 'https://example.com/.well-known/pay',
+  publicName: 'Payment Pointer',
```
Comment (wilsonianb) [inline packages/open-payments/src/test/helpers.ts:74]:
```suggestion
```
Could exclude the optional field by default

File: packages/open-payments/src/client/incoming-payment.ts (hunk 2 above)
Comment (mkurapov) [reply packages/open-payments/src/client/incoming-payment.ts:17]: Yup, I like that better

File: packages/open-payments/src/client/ilp-stream-connection.test.ts (hunk 1 above)
Comment (mkurapov) [reply packages/open-payments/src/client/ilp-stream-connection.test.ts:42]: It's already testing this (albeit indirectly): if the `createResponseValidator` doesn't get the proper `path` and `method` when, the test will fail based on the expectation of `true` on L57

File: packages/open-payments/src/client/index.ts [hunk 8]
```diff
@@ -41,6 +41,20 @@ export interface RouteDeps extends BaseDeps {
   logger: Logger
 }
 
+export interface UnauthenticatedResourceRequestArgs {
```
Comment (mkurapov) [inline packages/open-payments/src/client/index.ts:44]: it's a bit wordy given what this contains, but I like how the types read in the `grant.ts`, where the first method (to create a grant) has `UnauthenticatedResourceRequestArgs`, while all the other ones are `ResourceRequestArgs` with the explicit `accessToken`

File: packages/open-payments/src/client/ilp-stream-connection.test.ts (hunk 1 above)
Comment (wilsonianb) [reply packages/open-payments/src/client/ilp-stream-connection.test.ts:42]: Ah, the `true`! :+1:

Comment (mkurapov) [reply packages/open-payments/src/client/ilp-stream-connection.test.ts:42]: It's not the easiest to see when first reading it, but that was the best solution I could come up with to test the correct validator was passed in

File: packages/open-payments/src/client/index.ts [hunk 9]
```diff
@@ -41,6 +41,20 @@ export interface RouteDeps extends BaseDeps {
...
+
+export interface ResourceRequestArgs {
+  url: string
+  accessToken: string
+}
+
+export interface CollectionRequestArgs {
+  paymentPointer: string
+  accessToken: string
+}
```
Comment (wilsonianb) [inline packages/open-payments/src/client/index.ts:56]:
```suggestion
interface AuthenticatedRequestArgs {
  accessToken: string
}

export interface ResourceRequestArgs
  extends UnauthenticatedResourceRequestArgs,
    AuthenticatedRequestArgs {}

export interface CollectionRequestArgs extends AuthenticatedRequestArgs {
  paymentPointer: string
}
```

**Review by wilsonianb — DISMISSED** [review]

File: packages/open-payments/src/test/helpers.ts [hunk 10]
```diff
@@ -29,6 +31,25 @@ export const defaultAxiosInstance = createAxiosInstance({
...
+export const withEnvVariableOverride = (
+  override: Record<string, string>,
+  testCallback: () => Promise<void>
+): (() => Promise<void>) => {
+  return async () => {
+    const savedEnvVars = Object.assign({}, process.env)
+
+    for (const key in override) {
+      process.env[key] = override[key]
+    }
```
Comment (dragosp1011) [inline packages/open-payments/src/test/helpers.ts:43]:
```suggestion
    Object.assign(process.env, override)
```

**Review by wilsonianb — DISMISSED** [review]

File: packages/open-payments/src/client/quote.test.ts [hunk 11]
```diff
@@ -35,29 +35,6 @@ describe('quote', (): void => {
   const paymentPointer = 'http://localhost:1000/.well-known/pay'
   const accessToken = 'accessToken'
 
-  describe('createQuoteRoutes', (): void => {
```
Comment (mkurapov) [inline packages/open-payments/src/client/quote.test.ts:38]: already tested via `routes` describe block lower in the file

**Review by wilsonianb — APPROVED** [review]
